# Problem Set Week 1 MCB112 
The case of the dead sand mouse

## The Experiment
Moriarty's methods section says that he instantly kills a sand mouse at time t=0. At t=0, 12, 24, 48, and 96 hours, he dissects out the prefrontal cortex, prepared poly-A+ mRNA, and does RNA-seq. An example of the results is shown in Figure 1 from Moriarty et al.: arugula goes down after death as you might expect from being dead... but OSTC comes up at 24hrs then goes down, PRIM2 peaks at 48h, and kiwi is still coming up at 96h after death:

In his discussion, Moriarty says that these data provide evidence for an ancient program of cortical gene expression that causes the sand mouse's life to flash before its eyes (very, very slowly).

Well, maybe. But you suspect an artifact of some kind. You decide to look into the result. After carefully reading the paper, it looks like there aren't any obvious problems with the RNA-seq experiment itself. Each data point is an average over several dead sand mice, and the variation seems small, so it's not that this is just experimental noise. The dissections seem carefully done, so it's not that he mistakenly collected anatomically different chunks of brain (which could easily give gene expression differences). He also did some controls to make sure that the relative proportions of different cell types in the dissected tissue hadn't changed much (a common problem in differential RNA-seq analysis of tissues composed of mixtures of cell types: you measure a population average, so if cell type composition changes, it can look like gene expression changes.) Everything suggests to you that the expression level (TPM) calls are reproducible and robust, so if there's a problem, it's in his interpretation of the expression levels.

## The Data
The expression data for the paper are available in his Supplementary Data Table 1: Moriarty_SuppTable1.

It happens that you've also just been reading another paper from Irene Adler's lab that seems relevant. Adler et al. systematically measured mRNA synthesis rates (in mRNA/hr) and mRNA halflife (in hr), also in the prefrontal cortex, for every gene in the sand mouse. Those data are available in their Supplementary Data Table 2: Adler_SuppTable2.

Download both data files. 

### 1. Check that the gene names match
Write a script to compare the gene names in the two data files. Output the names that appear in Moriarty_SuppTable1 but not Adler_SuppTable2, if any. Use plain Python for this pset. Don't use any additional data wrangling packages, if you already happen to know it. One of the points of the pset is to be able to ingest crappily-formatted data files, by writing your own parsing code. 

In [5]:
sup_one = readlines(open("Sup_1.txt")) 

20032-element Vector{String}:
 "gene_name         0h    12h    24h    48h    96h "
 "anise              5.4    5.2    4.8    3.1    0.6 "
 "apricot           13.4   14.6   12.4    8.6    1.9 "
 "artichoke          4.2    1.5    0.5    0.0    0.0 "
 "arugula           66.9   23.9    7.0    0.5    0.0 "
 "asparagus         55.6   78.1  100.9  135.8  115.0 "
 "avocado           27.0   14.3    6.2    1.1    0.0 "
 "banana           101.3  132.2  126.7  130.7   68.7 "
 "basil              3.1    1.1    0.4    0.0    0.0 "
 "beet               3.8    1.1    0.3    0.0    0.0 "
 "blackberry        23.9   15.6    9.1    2.5    0.1 "
 "blueberry          7.3    4.5    2.0    0.4    0.0 "
 "broccoli          16.4   14.1   11.1    5.4    0.7 "
 ⋮
 "AL592183.1        33.7   45.8   52.5   62.8   53.5 "
 "AC007325.1        77.7   51.8   27.9    6.5    0.2 "
 "AC007325.4        34.3   34.4   38.6   30.0   12.6 "
 "AC007325.2         7.3    7.4    6.7    4.0    1.0 "
 "BX072566.1        26.9   29.1   

In [6]:
sup_two = readlines(open("Sup_2.txt"))

20032-element Vector{String}:
 "gene_name     synth_rate halflife"
 "CYB5R3               3.2     19.9"
 "MMAA                 1.1      9.5"
 "SLC24A1              0.8     14.8"
 "OR10D3              12.6     14.0"
 "IFITM5               2.5     20.3"
 "CDC5L                0.7      8.3"
 "DPYSL2               1.4     26.5"
 "ERCC6                1.0     11.3"
 "SYCE2                1.7     23.0"
 "CHST7                2.1      5.5"
 "PFDN2                6.6     17.2"
 "PHKB                 0.3     17.5"
 ⋮
 "FGF23                6.6      8.8"
 "SPINK6               1.9      9.3"
 "RUNX1T1              1.4     14.8"
 "GABRB1               3.4     23.6"
 "GJC1                 4.4     23.1"
 "TFRC                 2.4     99.5"
 "STK24                1.7      9.4"
 "MTUS1                4.7     21.0"
 "TRIM42               2.2     17.2"
 "IRF7                10.4     10.2"
 "HSPE1                2.8     17.0"
 "EYS                  1.8     14.1"

In [12]:
split(sup_one[2], " ")[1]

"anise"

In [8]:
sup_one[2]

"anise              5.4    5.2    4.8    3.1    0.6 "

In [13]:
sup_one_genes = []
for line in sup_one[2:end]
    split_line = split(line)
    gene_name = split_line[1]
    push!(sup_one_genes, gene_name)
end 

In [14]:
in(sup_one[1], sup_one)

true

In [15]:
sup_two_genes = []
for line in sup_two[2:end]
    split_line = split(line)
    gene_name = split_line[1]
    push!(sup_two_genes, gene_name)
end 

In [16]:
common_genes = intersect(sup_one_genes, sup_two_genes)

20003-element Vector{Any}:
 "anise"
 "apricot"
 "artichoke"
 "arugula"
 "asparagus"
 "avocado"
 "banana"
 "basil"
 "beet"
 "blackberry"
 "blueberry"
 "broccoli"
 "butternut"
 ⋮
 "AL592183.1"
 "AC007325.1"
 "AC007325.4"
 "AC007325.2"
 "BX072566.1"
 "AL354822.1"
 "AC023491.2"
 "AC004556.1"
 "AC233755.2"
 "AC233755.1"
 "AC240274.1"
 "AC213203.1"

In [17]:
length(sup_one_genes)

20031

In [18]:
length(sup_two_genes) 

20031

In [19]:
length(common_genes)

20003

Is there a difference? 
Yes, there are 28 genes that are not shared between the two supplementary tables

In [20]:
corrupt_genes = setdiff(sup_one_genes, sup_two_genes)

26-element Vector{Any}:
 "15-Sep"
 "2-Mar"
 "1-Mar"
 "10-Sep"
 "7-Mar"
 "4-Mar"
 "2-Sep"
 "11-Sep"
 "6-Mar"
 "11-Mar"
 "3-Mar"
 "8-Sep"
 "7-Sep"
 "14-Sep"
 "6-Sep"
 "1-Dec"
 "8-Mar"
 "5-Mar"
 "9-Mar"
 "12-Sep"
 "1-Sep"
 "4-Sep"
 "10-Mar"
 "9-Sep"
 "5-Sep"
 "3-Sep"

All of these different genes are dates - likely due to xlsx formating error. Omit them from the analysis

#### Create new supplementary files with these names ommitted 


In [21]:
new_sup_one = Vector()
@time for line in sup_one
        gene = split(line, " ")[1]
            if gene ∉ corrupt_genes
            push!(new_sup_one, line)        
            end
    end 

  0.164224 seconds (643.74 k allocations: 42.401 MiB, 28.75% compilation time)


In [22]:
length(new_sup_one)

20004

The incorrectly formatted genes are from supplementary file one. 

In [23]:
new_sup_two = Vector()
@time for line in sup_two
        gene = split(line, " ")[1]
            if gene ∉ corrupt_genes
            push!(new_sup_two, line)        
            end
    end 

  0.072579 seconds (640.52 k allocations: 42.186 MiB, 10.96% gc time)


In [24]:
length(new_sup_two)

20032

### 2. Explore the data 
Output the five genes with the highest mRNA synthesis rate

I am thinking that I will create a dict which maps the gene name to the synthesis rate, and then I can easily sort the synthesis values.

Because the supplementary file is inconsistently formatted, with varying amounts of whitespace delimiters between the variable values, we should tidy it up and make it into a consistent comma delimited array. 

In [45]:
# We don't want to print white spaces, but remove them! 
final_sup_two = Vector{String}()
for line in new_sup_two[2:end]
    new_line = Vector()
    for char in eachindex(line) # each eachindex() function to list out iterate over indices
        if line[char] != ' ' # if we encounter a non-space character, send it to the array
            new_line = push!(new_line, line[char]) 
        elseif line[char] == ' ' && line[char-1] != ' ' # If we encounter a space, but the previous index is not a space, send it to array
            new_line = push!(new_line, ',')
        end 
    end 
    push!(final_sup_two, join(new_line))
end       

In [26]:
sup_two_dict = Dict{AbstractString, Float64}() # Has to be this specific Type fields as otherwise sorting doesn't work as intended
for gene in final_sup_two[2:end]
    split_line = split(gene, ",")
    sup_two_dict[split_line[1]] = parse(Float64, split_line[2])
end 

In [27]:
sup_two_dict

Dict{AbstractString, Float64} with 20030 entries:
  "MMP11"     => 3.7
  "TMOD4"     => 14.1
  "ETV4"      => 0.8
  "DNASE2"    => 0.8
  "PRH1"      => 11.1
  "MOB4"      => 4.2
  "KAT2A"     => 3.0
  "GCNT3"     => 1.0
  "SREBF2"    => 5.9
  "APOC2"     => 2.6
  "CDC40"     => 2.0
  "HLA-B"     => 1.9
  "FAXC"      => 0.6
  "NAGS"      => 4.7
  "CYP2C19"   => 6.6
  "TMEM74B"   => 5.2
  "CNN3"      => 6.3
  "HOXD4"     => 1.5
  "TNFRSF11B" => 4.4
  "AMZ2"      => 7.3
  "ACER2"     => 6.3
  "TPP2"      => 1.1
  "THBS4"     => 7.0
  "MED9"      => 4.1
  "NBPF4"     => 3.3
  ⋮           => ⋮

In [28]:
sorted_sup_two_rates = sort(collect(sup_two_dict), by = x -> x[2], rev = true) ; first(sorted_sup_two_rates, 5)

5-element Vector{Pair{AbstractString, Float64}}:
 "CCDC169-SOHLH2" => 114.7
         "DDX60L" => 80.1
          "LRRK1" => 75.3
       "SLC25A45" => 70.1
          "FARP1" => 68.2

#### Answer
The five genes with the highest mRNA synthesis rate are   
1. CCDC169-SOHLH2 => 114.7
2. DDX60L => 80.1
3. LRRK1 => 75.3
4. SLC25A45 => 70.1
5. FARP1 => 68.2

Output the five genes with the longest mRNA halflife

In [29]:
sup_two_dict_halflife = Dict{AbstractString, Float64}() # Has to be this specific Type fields as otherwise sorting doesn't work as intended
for gene in final_sup_two[2:end]
    split_line = split(gene, ",")
    sup_two_dict_halflife[split_line[1]] = parse(Float64, split_line[3])
end 

In [30]:
sup_two_dict_halflife

Dict{AbstractString, Float64} with 20030 entries:
  "MMP11"     => 18.6
  "TMOD4"     => 6.5
  "ETV4"      => 9.2
  "DNASE2"    => 18.7
  "PRH1"      => 12.6
  "MOB4"      => 11.8
  "KAT2A"     => 13.7
  "GCNT3"     => 16.2
  "SREBF2"    => 22.6
  "APOC2"     => 20.9
  "CDC40"     => 16.5
  "HLA-B"     => 18.9
  "FAXC"      => 8.7
  "NAGS"      => 10.2
  "CYP2C19"   => 6.3
  "TMEM74B"   => 8.2
  "CNN3"      => 16.5
  "HOXD4"     => 16.6
  "TNFRSF11B" => 17.6
  "AMZ2"      => 13.7
  "ACER2"     => 16.8
  "TPP2"      => 13.0
  "THBS4"     => 6.6
  "MED9"      => 21.0
  "NBPF4"     => 18.0
  ⋮           => ⋮

In [31]:
sorted_sup_two_halflife = sort(collect(sup_two_dict_halflife), by = x -> x[2], rev = true) ; first(sorted_sup_two_halflife, 5)

5-element Vector{Pair{AbstractString, Float64}}:
   "TFRC" => 99.5
 "SPINK8" => 71.8
  "DIRC1" => 67.3
  "PLA1A" => 61.5
 "SAMSN1" => 58.1

#### Answer
The tope five genes with the longest half-life are
1. TFRC => 99.5
2. SPINK8 => 71.8
3. DIRC1 => 67.3
4. PLA1A => 61.5
5. SAMSN1 => 58.1

#### 3. Figure out what happened
Write a script that merges the two data files, line by line, merging them on gene name. That is, for each line in file 1 for gene X, find the corresponding line for gene X in file 2; we're going to write a single output file with one line per gene. The genes are in different orders in the files, so this merge isn't entirely trivial. For any gene name X that isn't found in both files (cough cough Excel corruption cough cough) just skip it. For each gene name that is found in both files, output one whitespace-delimited, column-justified data line consisting of 7 fields per line:

Create a final, cleanly formatted supplementary file 2

In [32]:
final_sup_one = []
for line in new_sup_one[2:end]
    new_line = Vector()
    for char in eachindex(line) # each eachindex() function to list out iterate over indices
        if line[char] != ' ' # if we encounter a non-space character, send it to the array
            new_line = push!(new_line, line[char]) 
        elseif line[char] == ' ' && line[char-1] != ' ' # If we encounter a space, but the previous index is not a space, send it to array
            new_line = push!(new_line, ',')
        end 
    end 
    push!(final_sup_one, join(new_line))
end   

In [49]:
merged_vector = Vector{String}()
for line in final_sup_one
    new_merged_line = Vector()
    split_line_measures = split(line, ',')[2:end]
    split_line_gene = split(line, ',')[1]
    findall(r"split_line_gene", final_sup_two)
    #if split_line_gene ∈ keys(sup_two_dict_halflife)
     #   push!(
    #end 
end 

LoadError: MethodError: no method matching String(::Vector{String})

Closest candidates are:
  String(::Core.Compiler.LazyString)
   @ Core strings/lazy.jl:80
  String(::Vector{UInt8})
   @ Base strings/string.jl:67
  String(::SubString{String})
   @ Base strings/substring.jl:66
  ...
